circuit/src/merkle_tree/native.rs

```diff
@@ -0,0 +1,155 @@
+use num_traits::pow;
+use std::collections::HashMap;
+
+use crate::{params::RoundParams, poseidon::native::Poseidon};
+const WIDTH: usize = 5;
+
+#[derive(Clone, Debug)]
+/// MerkleTree structure
+pub struct MerkleTree<F: FieldExt, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// HashMap to keep the level and index of the nodes
+	pub(crate) nodes: HashMap<usize, Vec<F>>,
+	/// Height of the tree
+	pub(crate) height: usize,
+	/// Root of the tree
+	pub(crate) root: F,
+	/// PhantomData for the params
+	_params: PhantomData<P>,
+}
+
+impl<F: FieldExt, P> MerkleTree<F, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Build a MerkleTree from given leaf nodes and height
+	pub fn build_tree(mut leaves: Vec<F>, height: usize) -> Self {
+		assert!(leaves.len() <= pow(2, height));
+		// 0th level is the leaf level and the max level is the root level
+		let mut nodes = HashMap::new();
+		// Assign zero to the leaf values if they are empty
+		for _i in leaves.len()..pow(2, height) {
+			leaves.push(F::zero());
+		}
+		nodes.insert(0, leaves);
+
+		for level in 0..height {
+			let mut hashes = Vec::new();
+			for i in 0..nodes[&level].len() {
+				if i % 2 != 0 {
+					continue;
+				}
+				let pos_inputs =
+					[nodes[&level][i], nodes[&level][i + 1], F::zero(), F::zero(), F::zero()];
+				let hasher: Poseidon<F, WIDTH, P> = Poseidon::new(pos_inputs);
+				hashes.push(hasher.permute()[0]);
+			}
+			nodes.insert(level + 1, hashes);
+		}
+		let root = nodes[&height][0].clone();
+		MerkleTree { nodes, height, root, _params: PhantomData }
+	}
+}
+
+#[derive(Clone)]
+/// Path structure
+pub struct Path<F: FieldExt, const ARITY: usize, const LENGTH: usize, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Value that is based on for construction of the path
+	pub(crate) value: F,
+	/// Array that keeps the path
+	pub(crate) path_arr: [[F; ARITY]; LENGTH],
+	/// PhantomData for the params
+	_params: PhantomData<P>,
+}
+
+impl<F: FieldExt, const ARITY: usize, const LENGTH: usize, P> Path<F, ARITY, LENGTH, P>
+where
+	P: RoundParams<F, WIDTH>,
+{
+	/// Find path for the given value to the root
+	pub fn find_path(merkle_tree: &MerkleTree<F, P>, value: F) -> Path<F, ARITY, LENGTH, P> {
+		//
+		// TODO: This way of finding index will fail if we have same inputs
+		//
+		let mut value_index = merkle_tree.nodes[&0].iter().position(|x| x == &value).unwrap();
+		let mut path_arr: [[F; ARITY]; LENGTH] = [[F::zero(); ARITY]; LENGTH];
+		// Childs for a parent node is 2n and 2n + 1.
+		// value_index keeps index of that nodes in reverse order to apply this
+		// algorithm.
+		for level in 0..merkle_tree.height {
+			if value_index % 2 == 1 {
+				path_arr[level][0] = merkle_tree.nodes[&level][value_index - 1];
+				path_arr[level][1] = merkle_tree.nodes[&level][value_index];
+			} else {
+				path_arr[level][0] = merkle_tree.nodes[&level][value_index];
+				path_arr[level][1] = merkle_tree.nodes[&level][value_index + 1]
+			}
+			value_index = value_index / 2;
```

Comment:
`ARITY` stands for the number of nodes to be hashed together at each level. In this part we are ignoring the `ARITY` constant and treating this tree as binary (2 nodes at each level).
You can either remove ARITY constant and replace it with 2, or handle the logic.

Reply:
I was trying to do some logic to keep last variable (root) as a single element and I miss to change that back to 2. Fixing it immediately and doing last checks for the comment lines.